File: common/modules/debrid/premiumize.js
// relative import keeps this module loadable under plain Node for API tests
import DebridService, { DebridError, DebridAuthError, DebridNotCachedError, DebridUnavailableError } from './service.js'
import { Availability } from './availability.js'
import Debug from 'debug'
const debug = Debug('ui:debrid')

const API = 'https://www.premiumize.me/api'

// error codes worth explaining, anything else falls back to the API's own message
const errorMessages = {
  authentication_failed: 'Invalid Premiumize API key',
  permission_denied: 'Premiumize denied the request, check the account',
  account_limit_reached: 'This Premiumize account has used up its fair use points or active jobs',
  service_limit_reached: 'This Premiumize account has reached its limit for this source',
  rate_limit_reached: 'Premiumize is rate limiting this account, try again shortly',
  service_down: 'Premiumize cannot reach this source right now',
  service_unsupported: 'Premiumize cannot process this kind of source',
  link_generation_failed: 'Premiumize could not generate a stream link, try again shortly'
}
// only these mean the key or account is the problem, the rest are per-request
const authCodes = ['authentication_failed', 'permission_denied']
// the account cannot take more work right now, rather than this release being a problem
const throttleCodes = ['rate_limit_reached', 'account_limit_reached', 'service_limit_reached']
// the same request will keep failing, so this release is not one Premiumize can serve
const deadCodes = ['service_unsupported', 'permanent_error']

/**
 * Premiumize implementation, see https://www.premiumize.me/api
 *
 * The easiest service to support, having kept the endpoints the others dropped: `/cache/check`
 * answers a whole results list for free, and `/transfer/directdl` returns every stream link for a
 * magnet in one call without storing anything, so there is no add or cleanup path at all.
 *
 * `/transfer/list` never says which info hash a transfer came from, so badges come from the cache
 * endpoint alone.
 */
export default class Premiumize extends DebridService {
  static id = 'premiumize'
  static title = 'Premiumize'
  static available = true
  // a real cache endpoint, so badges cost one request for the whole results list
  static availabilityCheck = 'batch'
  static maxBatch = 100
  static limits = { maxConcurrent: 3, minTime: 250 } // no documented allowance, so be modest

  /** Failures arrive inside a 200; the payload is top level rather than in an envelope. */
  unwrap (json) {
    if (!json || typeof json !== 'object' || !('status' in json)) return json
    if (json.status === 'error') throw this.mapError(200, json)
    return json
  }

  mapError (status, json) {
    const code = json?.code
    const message = errorMessages[code] || json?.message || `Request failed with status ${status}`
    if (authCodes.includes(code) || ((status === 401 || status === 403) && !code)) return new DebridAuthError(message, { status, code })
    return new DebridError(message, { status, code })
  }

  /** @param {any} error */
  throttled (error) {
    return super.throttled(error) || throttleCodes.includes(error?.code)
  }

  async validate () {
    const account = await this.request(`${API}/account/info`)
    // free accounts stream cached content through the same endpoints, so premium is not required
    if (!account?.customer_id) throw new DebridAuthError('Premiumize did not recognise this API key')
    return { username: `Premiumize ${account.customer_id}`, expires: account.premium_until ? new Date(account.premium_until * 1_000).toISOString() : undefined }
  }

  /** Nothing to read: transfers carry no info hash. The cache endpoint covers this instead. */
  async listAvailability () {
    return new Map()
  }

  /** @see listAvailability - nothing reads the account listing. */
  async fetchListing () {
    return []
  }

  /**
   * One request, however many releases, and it costs the account nothing.
   * @param {string[]} hashes
   */
  async checkAvailabilityBatch (hashes) {
    const checked = await this.request(`${API}/cache/check`, { method: 'POST', body: { 'items[]': hashes.map(hash => Premiumize.toMagnet(hash)) } })
    const cached = checked?.response || [] // parallel arrays indexed by request order, not keyed by hash
    return new Map(hashes.map((hash, index) => [hash, cached[index] ? Availability.CACHED : Availability.AVAILABLE]))
  }

  async resolve (magnet, { fileFilter = () => true, pickFile, maxFiles = this.config.maxFiles } = {}) {
    const hash = Premiumize.parseHash(magnet)
    const magnetURI = Premiumize.toMagnet(magnet)
    if (!magnetURI) throw new DebridError('Premiumize needs a magnet link or info hash to resolve')
    const content = await this.#directdl(magnetURI)
    const wanted = content
      .map(entry => ({ name: filePath(entry).split('/').pop(), path: filePath(entry), size: Number(entry.size) || 0, url: entry.link }))
      .filter(file => fileFilter(file.path))
    if (!wanted.length) throw new DebridError('No playable files in this torrent')
    const target = pickFile ? await pickFile(wanted) : [...wanted].sort((a, b) => b.size - a.size)[0]
    const files = Premiumize.windowFiles(wanted, target, maxFiles)
    const name = torrentName(files)
    debug(`Resolved ${files.length} files for ${name}`)
    return { hash, name, files }
  }

  /**
   * Every stream link for a magnet, read out of the cache in one call. Touches nothing, so a
   * miss comes back empty or as a code rather than needing a check first.
   * @param {string} magnetURI
   * @returns {Promise<any[]>} The transfer's file entries.
   */
  async #directdl (magnetURI) {
    let transfer = null
    try {
      transfer = await this.request(`${API}/transfer/directdl`, { method: 'POST', body: { src: magnetURI } })
    } catch (error) {
      // the API groups its codes by whether the same request could ever succeed, which maps
      // straight onto what playback needs to know
      if (deadCodes.includes(error?.code)) throw new DebridUnavailableError(error.message, { status: error.status, code: error.code })
      if (error?.code === 'not_found') throw new DebridNotCachedError() // it can still fetch it, just not now
      throw error
    }
    const content = (transfer?.content || []).filter(entry => entry?.link)
    if (!content.length) throw new DebridNotCachedError() // directdl only reads the cache
    return content
  }
}

/**
 * Paths arrive without a leading slash; Shiru's file objects are rooted like the torrent client's.
 * @param {any} entry
 */
function filePath (entry) {
  const path = entry?.path || ''
  return path.startsWith('/') ? path : `/${path}`
}

/**
 * A name for the release, which directdl never states: the folder a pack sits under, or the file.
 * @param {{ path: string, name: string }[]} files
 */
function torrentName (files) {
  const [first] = files
  const folder = first.path.split('/')[1]
  return files.every(file => file.path.startsWith(`/${folder}/`)) ? folder : first.name
}

File: app/lib/infrastructure/debrid/providers/premiumize.dart
import '../../../domain/models/availability.dart';
import '../../network/transport.dart';
import '../client.dart';
import '../errors.dart';
import '../hash.dart';
import '../window.dart';
import 'json.dart';
import 'provider.dart';

const _api = 'https://www.premiumize.me/api';

const _knownMessages = <String, String>{
  'authentication_failed': 'Invalid Premiumize API key',
  'permission_denied': 'Premiumize denied the request, check the account',
  'account_limit_reached':
      'This Premiumize account has used up its fair use points or active jobs',
  'service_limit_reached':
      'This Premiumize account has reached its limit for this source',
  'rate_limit_reached':
      'Premiumize is rate limiting this account, try again shortly',
  'service_down': 'Premiumize cannot reach this source right now',
  'service_unsupported': 'Premiumize cannot process this kind of source',
  'link_generation_failed':
      'Premiumize could not generate a stream link, try again shortly',
};

const _authCodes = {'authentication_failed', 'permission_denied'};
const _throttleCodes = {
  'rate_limit_reached',
  'account_limit_reached',
  'service_limit_reached',
};
const _deadCodes = {'service_unsupported', 'permanent_error'};

class _PremiumizeDialect extends Dialect {
  const _PremiumizeDialect();

  @override
  Object? unwrap(Object? json) {
    if (field(json, 'status') == 'error') throw mapError(200, json);
    return json;
  }

  @override
  DebridFailure mapError(int status, Object? json) {
    final code = jsonString(field(json, 'code'));
    final message =
        _knownMessages[code] ??
        safeServiceMessage(
          jsonString(field(json, 'message')),
          'Request failed with status $status',
        );
    final auth =
        (code != null && _authCodes.contains(code)) ||
        ((status == 401 || status == 403) && code == null);
    return auth
        ? DebridFailure.auth(message, status: status, code: code)
        : DebridFailure.service(message, status: status, code: code);
  }
}

class PremiumizeProvider implements DebridProvider {
  PremiumizeProvider(String apiKey, HttpTransport transport, Clock clock)
    : client = DebridApiClient(providerConfig, apiKey, transport, clock);

  static const providerConfig = ProviderConfig(
    id: 'premiumize',
    title: 'Premiumize',
    auth: AuthScheme.bearer,
    authParam: 'apikey',
    encoding: BodyEncoding.form,
    nominalLatencyMs: 300,
    maxFiles: 60,
    availabilityCheck: AvailabilityCheck.batch,
    checkAddsMagnets: false,
    maxBatch: 100,
    maxProbes: 10,
    maxConcurrent: 3,
    minTimeMs: 250,
  );

  static const _dialect = _PremiumizeDialect();

  @override
  final DebridApiClient client;

  @override
  ProviderConfig get config => providerConfig;

  @override
  bool throttled(DebridFailure error) {
    final code = _errorCode(error);
    return error.throttled || (code != null && _throttleCodes.contains(code));
  }

  Future<List<Object?>> _directDl(String magnet, {CancelToken? cancel}) async {
    Object? transfer;
    try {
      transfer = await client.request(
        _dialect,
        '$_api/transfer/directdl',
        RequestOpts(method: HttpMethod.post, body: [('src', magnet)]),
        cancel: cancel,
      );
    } on DebridFailure catch (error) {
      final code = _errorCode(error);
      if (code != null && _deadCodes.contains(code)) {
        throw DebridFailure.unavailable(error.message);
      }
      if (code == 'not_found') throw const DebridFailure.notCached();
      rethrow;
    }
    final content = [
      for (final entry in jsonList(field(transfer, 'content')))
        if (nonEmptyString(field(entry, 'link')) != null) entry,
    ];
    if (content.isEmpty) throw const DebridFailure.notCached();
    return content;
  }

  @override
  Future<AccountInfo> validate({CancelToken? cancel}) async {
    final account = await client.request(
      _dialect,
      '$_api/account/info',
      const RequestOpts(),
      cancel: cancel,
    );
    final customer = switch (field(account, 'customer_id')) {
      String value when value.isNotEmpty => value,
      num value when value != 0 => value.toString(),
      _ => null,
    };
    if (customer == null) {
      throw const DebridFailure.auth(
        'Premiumize did not recognise this API key',
      );
    }
    return AccountInfo(
      username: 'Premiumize $customer',
      expires: isoFromUnixSeconds(field(account, 'premium_until')),
    );
  }

  @override
  Future<Map<String, Availability>> listAvailability({
    CancelToken? cancel,
  }) async => const {};

  @override
  Future<Map<String, Availability>> checkAvailabilityBatch(
    List<String> hashes, {
    CancelToken? cancel,
  }) async {
    final items = [for (final hash in hashes) toMagnet(hash) ?? ''];
    final checked = await client.request(
      _dialect,
      '$_api/cache/check',
      RequestOpts(method: HttpMethod.post, body: [('items[]', items)]),
      cancel: cancel,
    );
    final response = jsonList(field(checked, 'response'));
    return {
      for (var index = 0; index < hashes.length; index++)
        hashes[index]: index < response.length && jsonTruthy(response[index])
            ? Availability.cached
            : Availability.available,
    };
  }

  @override
  Future<Availability> probeAvailability(String hash, {CancelToken? cancel}) =>
      Future.error(
        const DebridFailure.service(
          'Premiumize answers availability in batches via /cache/check',
        ),
      );

  @override
  Future<ResolvedFiles> resolve(
    String magnet,
    ResolveOptions opts, {
    CancelToken? cancel,
  }) async {
    final hash = parseHash(magnet) ?? '';
    final magnetUri = toMagnet(magnet);
    if (magnetUri == null) {
      throw const DebridFailure.service(
        'Premiumize needs a magnet link or info hash to resolve',
      );
    }
    final content = await _directDl(magnetUri, cancel: cancel);
    final wanted = <DebridFileInfo>[
      for (final entry in content)
        if (opts.fileFilter == null || opts.fileFilter!(_entryPath(entry)))
          DebridFileInfo(
            name: basename(_entryPath(entry)),
            path: _entryPath(entry),
            size: jsonInt(field(entry, 'size')) ?? 0,
            url: nonEmptyString(field(entry, 'link')) ?? '',
          ),
    ];
    if (wanted.isEmpty) {
      throw const DebridFailure.service('No playable files in this torrent');
    }
    final target = opts.pickFile == null
        ? _largestIndex(wanted)
        : opts.pickFile!([
            for (var index = 0; index < wanted.length; index++)
              (index, wanted[index].path, wanted[index].size),
          ]);
    final maxFiles = opts.maxFiles ?? config.maxFiles;
    final targetPath = target == null || target >= wanted.length
        ? null
        : wanted[target].path;
    final files = secureFiles(
      windowFiles(wanted, target, maxFiles),
      config.title,
    );
    return ResolvedFiles(
      hash: hash,
      name: _torrentName(files),
      files: files,
      targetPath: targetPath,
    );
  }

  @override
  Future<void> retryCleanup() async {}
}

String? _errorCode(DebridFailure error) => error.code;

String _entryPath(Object? entry) =>
    rootedPath(jsonString(field(entry, 'path')) ?? '');

int _largestIndex(List<DebridFileInfo> files) {
  var best = 0;
  for (var index = 1; index < files.length; index++) {
    if (files[index].size > files[best].size) best = index;
  }
  return best;
}

String _torrentName(List<DebridFileInfo> files) {
  final first = files.first;
  final parts = first.path.split('/');
  final folder = parts.length > 1 ? parts[1] : '';
  final prefix = '/$folder/';
  return folder.isNotEmpty &&
          files.every((file) => file.path.startsWith(prefix))
      ? folder
      : first.name;
}
